File: python/flyql/core/key.py
from dataclasses import dataclass, field
from typing import Any, List, Optional, Tuple

from flyql.core.exceptions import FlyqlError, KeyParseError
from flyql.core.range import Range


@dataclass
class KeyTransformer:
    """Parsed transformer invocation from a key pipeline (e.g., ``upper`` or
    ``format("YYYY")``). Carries the transformer's source ranges so tooling
    can map the AST back to the raw input.
    """

    name: str
    arguments: List[Any]
    range: Range
    name_range: Range
    argument_ranges: List[Range] = field(default_factory=list)


class Key:
    def __init__(
        self,
        segments: List[str],
        raw: Optional[str] = None,
        quoted_segments: Optional[List[bool]] = None,
        transformers: Optional[List[KeyTransformer]] = None,
        range: Optional[Range] = None,
        segment_ranges: Optional[List[Range]] = None,
    ):
        self.segments = segments
        self.is_segmented = len(segments) > 1
        self.raw = raw if raw is not None else ".".join(segments)
        self.quoted_segments = (
            quoted_segments if quoted_segments is not None else [False] * len(segments)
        )
        self.transformers: List[KeyTransformer] = (
            transformers if transformers is not None else []
        )
        if range is None:
            # Synthesize a default range from the raw string length for
            # direct SDK construction. The parser always passes real ranges.
            range = Range(0, len(self.raw))
        if segment_ranges is None:
            segment_ranges = []
            off = 0
            for i, seg in enumerate(segments):
                start = off
                end = off + len(seg)
                segment_ranges.append(Range(start, end))
                off = end + 1  # account for '.' separator
        self.range = range
        self.segment_ranges = segment_ranges


class KeyParser:
    def __init__(self) -> None:
        self.input = ""
        self.pos = 0
        self.base_offset = 0
        self.segments: List[str] = []
        self.quoted_segments: List[bool] = []
        self.segment_ranges: List[Range] = []
        self.current_segment = ""
        self.current_segment_quoted = False
        self.current_segment_has_content = False
        self.current_segment_start = -1

    def peek(self, offset: int = 0) -> Optional[str]:
        pos = self.pos + offset
        return self.input[pos] if pos < len(self.input) else None

    def advance(self) -> Optional[str]:
        char = self.peek()
        self.pos += 1
        return char

    def parse_escape_sequence(self) -> str:
        self.advance()  # Skip backslash
        char = self.peek()

        if char == "'":
            self.advance()
            return "'"
        elif char == '"':
            self.advance()
            return '"'
        elif char == "\\":
            self.advance()
            return "\\"
        elif char == "n":
            self.advance()
            return "\n"
        elif char == "t":
            self.advance()
            return "\t"
        elif char is not None:
            result = char
            self.advance()
            return result
        else:
            raise KeyParseError(
                f"Key parsing error: Incomplete escape sequence at position {self.pos}",
                range=Range(
                    self.base_offset + self.pos,
                    self.base_offset + self.pos + 1,
                ),
            )

    def parse_quoted_segment(self, quote_char: str) -> None:
        # If no content yet, segment starts at the opening quote position.
        if not self.current_segment_has_content and self.current_segment_start == -1:
            self.current_segment_start = self.pos
        self.advance()  # Skip opening quote

        while self.peek() is not None:
            char = self.peek()

            if char == "\\":
                self.current_segment += self.parse_escape_sequence()
            elif char == quote_char:
                self.advance()  # Skip closing quote
                return
            else:
                char = self.advance()
                if char is not None:
                    self.current_segment += char

        raise KeyParseError(
            f"Key parsing error: Unterminated quoted segment starting at position {self.pos}",
            range=Range(
                self.base_offset + self.pos,
                self.base_offset + self.pos,
            ),
        )

    def parse_normal_segment(self) -> None:
        while self.peek() is not None:
            char = self.peek()

            if char == ".":
                return
            elif char == "'":
                is_first = not self.current_segment_has_content
                self.parse_quoted_segment("'")
                if is_first:
                    self.current_segment_quoted = True
                self.current_segment_has_content = True
            elif char == '"':
                is_first = not self.current_segment_has_content
                self.parse_quoted_segment('"')
                if is_first:
                    self.current_segment_quoted = True
                self.current_segment_has_content = True
            elif char == "\\":
                if self.current_segment_start == -1:
                    self.current_segment_start = self.pos
                self.current_segment += self.parse_escape_sequence()
                self.current_segment_has_content = True
            else:
                if self.current_segment_start == -1:
                    self.current_segment_start = self.pos
                char = self.advance()
                if char is not None:
                    self.current_segment += char
                self.current_segment_has_content = True

    def parse(self, key_string: str, base_offset: int = 0) -> Key:
        self.input = key_string
        self.pos = 0
        self.base_offset = base_offset
        self.segments = []
        self.quoted_segments = []
        self.segment_ranges = []
        self.current_segment = ""
        self.current_segment_quoted = False
        self.current_segment_has_content = False
        self.current_segment_start = -1

        key_range = Range(base_offset, base_offset + len(key_string))

        if not self.input:
            return Key([], self.input, [], range=key_range, segment_ranges=[])

        while self.pos < len(self.input):
            seg_start_before = self.pos
            self.parse_normal_segment()
            seg_end = self.pos

            # If segment was empty (e.g. ".foo" where leading dot produced empty seg),
            # the segment's range is the zero-width span at the current position.
            if self.current_segment_start == -1:
                seg_start = self.base_offset + seg_start_before
            else:
                seg_start = self.base_offset + self.current_segment_start
            self.segments.append(self.current_segment)
            self.quoted_segments.append(self.current_segment_quoted)
            self.segment_ranges.append(Range(seg_start, self.base_offset + seg_end))
            self.current_segment = ""
            self.current_segment_quoted = False
            self.current_segment_has_content = False
            self.current_segment_start = -1

            if self.peek() == ".":
                self.advance()  # Skip dot
                if self.pos >= len(self.input):
                    self.segments.append("")
                    self.quoted_segments.append(False)
                    self.segment_ranges.append(
                        Range(
                            self.base_offset + self.pos,
                            self.base_offset + self.pos,
                        )
                    )

        return Key(
            self.segments,
            self.input,
            self.quoted_segments,
            range=key_range,
            segment_ranges=self.segment_ranges,
        )


def _parse_transformer_arguments(
    args_str: str, base_offset: int
) -> Tuple[List[Any], List[Range]]:
    args: List[Any] = []
    ranges: List[Range] = []
    i = 0
    while i < len(args_str):
        # Skip leading whitespace BEFORE capturing the argument start.
        while i < len(args_str) and args_str[i] == " ":
            i += 1
        if i >= len(args_str):
            break
        arg_start = i
        if args_str[i] in ('"', "'"):
            quote = args_str[i]
            i += 1
            val = ""
            while i < len(args_str) and args_str[i] != quote:
                if args_str[i] == "\\" and i + 1 < len(args_str):
                    i += 1
                    if args_str[i] == "t":
                        val += "\t"
                    elif args_str[i] == "n":
                        val += "\n"
                    else:
                        val += args_str[i]
                else:
                    val += args_str[i]
                i += 1
            if i < len(args_str):
                i += 1  # consume closing quote
            else:
                raise KeyParseError(
                    "unclosed string in transformer arguments",
                    Range(base_offset + arg_start, base_offset + i),
                )
            arg_end = i
            args.append(val)
            ranges.append(Range(base_offset + arg_start, base_offset + arg_end))
        else:
            val = ""
            while i < len(args_str) and args_str[i] not in (",", " "):
                val += args_str[i]
                i += 1
            arg_end = i
            try:
                args.append(int(val))
            except ValueError:
                try:
                    args.append(float(val))
                except ValueError:
                    args.append(val)
            ranges.append(Range(base_offset + arg_start, base_offset + arg_end))
        while i < len(args_str) and args_str[i] in (" ", ","):
            i += 1
    return args, ranges


def _parse_transformer_spec(spec: str, base_offset: int) -> KeyTransformer:
    paren_index = spec.find("(")
    if paren_index == -1:
        return KeyTransformer(
            name=spec,
            arguments=[],
            range=Range(base_offset, base_offset + len(spec)),
            name_range=Range(base_offset, base_offset + len(spec)),
            argument_ranges=[],
        )
    name = spec[:paren_index]
    close_index = spec.rfind(")")
    if close_index == -1:
        partial_args_str = spec[paren_index + 1 :]
        if partial_args_str:
            arg_values, arg_ranges = _parse_transformer_arguments(
                partial_args_str, base_offset + paren_index + 1
            )
            return KeyTransformer(
                name=name,
                arguments=arg_values,
                range=Range(base_offset, base_offset + len(spec)),
                name_range=Range(base_offset, base_offset + paren_index),
                argument_ranges=arg_ranges,
            )
        return KeyTransformer(
            name=name,
            arguments=[],
            range=Range(base_offset, base_offset + len(spec)),
            name_range=Range(base_offset, base_offset + paren_index),
            argument_ranges=[],
        )
    args_str = spec[paren_index + 1 : close_index]
    arg_values, arg_ranges = _parse_transformer_arguments(
        args_str, base_offset + paren_index + 1
    )
    return KeyTransformer(
        name=name,
        arguments=arg_values,
        range=Range(base_offset, base_offset + len(spec)),
        name_range=Range(base_offset, base_offset + paren_index),
        argument_ranges=arg_ranges,
    )


def parse_key(key_string: str, base_offset: int = 0) -> Key:
    parts = key_string.split("|")
    base_key_string = parts[0]
    transformer_specs = parts[1:] if len(parts) > 1 else []

    parser = KeyParser()
    key = parser.parse(base_key_string, base_offset)

    if transformer_specs:
        transformers: List[KeyTransformer] = []
        # First transformer spec starts after base key + '|' char.
        running_offset = base_offset + len(base_key_string) + 1
        for spec in transformer_specs:
            parsed = _parse_transformer_spec(spec, running_offset)
            if not parsed.name:
                raise KeyParseError(
                    "empty transformer name in key",
                    range=Range(running_offset, running_offset + len(spec)),
                )
            transformers.append(parsed)
            running_offset += len(spec) + 1
        key.transformers = transformers
        key.raw = key_string
        # Key.range covers entire pipeline including transformers.
        key.range = Range(base_offset, base_offset + len(key_string))

    return key


def __getattr__(name: str) -> Any:
    if name == "Transformer":
        import warnings

        warnings.warn(
            "flyql.core.key.Transformer is renamed to KeyTransformer and will be removed in 1.1.0. "
            "Update imports to `from flyql.core.key import KeyTransformer`.",
            DeprecationWarning,
            stacklevel=2,
        )
        return KeyTransformer
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")

File: javascript/src/core/key.js
import { KeyParseError } from './exceptions.js'
import { Range } from './range.js'

/**
 * Transformer — a parsed transformer invocation from a key pipeline
 * (e.g. upper or format("YYYY")). Carries source ranges so tooling
 * can map the AST back to the raw input.
 */
export class Transformer {
    constructor(name, args, range, nameRange, argumentRanges = []) {
        this.name = name
        this.arguments = args
        this.range = range
        this.nameRange = nameRange
        this.argumentRanges = argumentRanges
    }
}

export class Key {
    constructor(segments, raw = null, transformers = [], quotedSegments = null, range = null, segmentRanges = null) {
        this.segments = segments
        this.isSegmented = segments.length > 1
        this.raw = raw !== null ? raw : segments.join('.')
        this.transformers = transformers
        this.quotedSegments = quotedSegments !== null ? quotedSegments : segments.map(() => false)
        if (range === null) {
            range = new Range(0, this.raw.length)
        }
        if (segmentRanges === null) {
            segmentRanges = []
            let off = 0
            for (const seg of segments) {
                segmentRanges.push(new Range(off, off + seg.length))
                off += seg.length + 1
            }
        }
        this.range = range
        this.segmentRanges = segmentRanges
    }
}

export class KeyParser {
    constructor() {
        this.input = ''
        this.pos = 0
        this.baseOffset = 0
        this.segments = []
        this.quotedSegments = []
        this.segmentRanges = []
        this.currentSegment = ''
        this.currentSegmentQuoted = false
        this.currentSegmentHasContent = false
        this.currentSegmentStart = -1
    }

    peek(offset = 0) {
        const pos = this.pos + offset
        return pos < this.input.length ? this.input[pos] : null
    }

    advance() {
        const char = this.peek()
        this.pos += 1
        return char
    }

    parseEscapeSequence() {
        this.advance() // Skip backslash
        const char = this.peek()

        if (char === "'") {
            this.advance()
            return "'"
        } else if (char === '"') {
            this.advance()
            return '"'
        } else if (char === '\\') {
            this.advance()
            return '\\'
        } else if (char === 'n') {
            this.advance()
            return '\n'
        } else if (char === 't') {
            this.advance()
            return '\t'
        } else if (char !== null) {
            const result = char
            this.advance()
            return result
        } else {
            throw new KeyParseError(
                `Key parsing error: Incomplete escape sequence at position ${this.pos}`,
                new Range(this.baseOffset + this.pos, this.baseOffset + this.pos + 1),
            )
        }
    }

    parseQuotedSegment(quoteChar) {
        if (!this.currentSegmentHasContent && this.currentSegmentStart === -1) {
            this.currentSegmentStart = this.pos
        }
        this.advance() // Skip opening quote
        this.currentSegmentQuoted = true

        while (this.peek() !== null) {
            const char = this.peek()

            if (char === '\\') {
                this.currentSegment += this.parseEscapeSequence()
            } else if (char === quoteChar) {
                this.advance()
                return
            } else {
                const c = this.advance()
                if (c !== null) {
                    this.currentSegment += c
                }
            }
        }

        throw new KeyParseError(
            `Key parsing error: Unterminated quoted segment starting at position ${this.pos}`,
            new Range(this.baseOffset + this.pos, this.baseOffset + this.pos),
        )
    }

    parseNormalSegment() {
        while (this.peek() !== null) {
            const char = this.peek()

            if (char === '.') {
                return
            } else if (char === "'") {
                this.parseQuotedSegment("'")
                this.currentSegmentHasContent = true
            } else if (char === '"') {
                this.parseQuotedSegment('"')
                this.currentSegmentHasContent = true
            } else if (char === '\\') {
                if (this.currentSegmentStart === -1) {
                    this.currentSegmentStart = this.pos
                }
                this.currentSegment += this.parseEscapeSequence()
                this.currentSegmentHasContent = true
            } else {
                if (this.currentSegmentStart === -1) {
                    this.currentSegmentStart = this.pos
                }
                const c = this.advance()
                if (c !== null) {
                    this.currentSegment += c
                }
                this.currentSegmentHasContent = true
            }
        }
    }

    parse(keyString, baseOffset = 0) {
        this.input = keyString
        this.pos = 0
        this.baseOffset = baseOffset
        this.segments = []
        this.quotedSegments = []
        this.segmentRanges = []
        this.currentSegment = ''
        this.currentSegmentQuoted = false
        this.currentSegmentHasContent = false
        this.currentSegmentStart = -1

        const keyRange = new Range(baseOffset, baseOffset + keyString.length)

        if (!this.input) {
            return new Key([], this.input, [], [], keyRange, [])
        }

        while (this.pos < this.input.length) {
            const segStartBefore = this.pos
            this.parseNormalSegment()
            const segEnd = this.pos

            let segStart
            if (this.currentSegmentStart === -1) {
                segStart = this.baseOffset + segStartBefore
            } else {
                segStart = this.baseOffset + this.currentSegmentStart
            }

            this.segments.push(this.currentSegment)
            this.quotedSegments.push(this.currentSegmentQuoted)
            this.segmentRanges.push(new Range(segStart, this.baseOffset + segEnd))
            this.currentSegment = ''
            this.currentSegmentQuoted = false
            this.currentSegmentHasContent = false
            this.currentSegmentStart = -1

            if (this.peek() === '.') {
                this.advance()
                if (this.pos >= this.input.length) {
                    this.segments.push('')
                    this.quotedSegments.push(false)
                    this.segmentRanges.push(new Range(this.baseOffset + this.pos, this.baseOffset + this.pos))
                }
            }
        }

        return new Key(this.segments, this.input, [], this.quotedSegments, keyRange, this.segmentRanges)
    }
}

function parseTransformerArguments(argsStr, baseOffset) {
    const args = []
    const ranges = []
    let i = 0
    while (i < argsStr.length) {
        while (i < argsStr.length && argsStr[i] === ' ') i++
        if (i >= argsStr.length) break

        const argStart = i
        if (argsStr[i] === '"' || argsStr[i] === "'") {
            const quote = argsStr[i]
            i++
            let val = ''
            while (i < argsStr.length && argsStr[i] !== quote) {
                if (argsStr[i] === '\\' && i + 1 < argsStr.length) {
                    i++
                    if (argsStr[i] === 't') val += '\t'
                    else if (argsStr[i] === 'n') val += '\n'
                    else val += argsStr[i]
                } else {
                    val += argsStr[i]
                }
                i++
            }
            if (i < argsStr.length) {
                i++ // skip closing quote
            } else {
                throw new KeyParseError(
                    'unclosed string in transformer arguments',
                    new Range(baseOffset + argStart, baseOffset + i),
                )
            }
            const argEnd = i
            args.push(val)
            ranges.push(new Range(baseOffset + argStart, baseOffset + argEnd))
        } else {
            let val = ''
            while (i < argsStr.length && argsStr[i] !== ',' && argsStr[i] !== ' ') {
                val += argsStr[i]
                i++
            }
            const argEnd = i
            const num = Number(val)
            args.push(val !== '' && !isNaN(num) ? num : val)
            ranges.push(new Range(baseOffset + argStart, baseOffset + argEnd))
        }
        while (i < argsStr.length && (argsStr[i] === ' ' || argsStr[i] === ',')) i++
    }
    return { args, ranges }
}

function parseTransformerSpec(spec, baseOffset) {
    const parenIndex = spec.indexOf('(')
    if (parenIndex === -1) {
        return new Transformer(
            spec,
            [],
            new Range(baseOffset, baseOffset + spec.length),
            new Range(baseOffset, baseOffset + spec.length),
            [],
        )
    }
    const name = spec.substring(0, parenIndex)
    const closeIndex = spec.lastIndexOf(')')
    if (closeIndex === -1) {
        const partialArgsStr = spec.substring(parenIndex + 1)
        const { args, ranges } =
            partialArgsStr.length > 0
                ? parseTransformerArguments(partialArgsStr, baseOffset + parenIndex + 1)
                : { args: [], ranges: [] }
        return new Transformer(
            name,
            args,
            new Range(baseOffset, baseOffset + spec.length),
            new Range(baseOffset, baseOffset + parenIndex),
            ranges,
        )
    }
    const argsStr = spec.substring(parenIndex + 1, closeIndex)
    const { args, ranges } = parseTransformerArguments(argsStr, baseOffset + parenIndex + 1)
    return new Transformer(
        name,
        args,
        new Range(baseOffset, baseOffset + spec.length),
        new Range(baseOffset, baseOffset + parenIndex),
        ranges,
    )
}

export function parseKey(keyString, baseOffset = 0) {
    const parts = keyString.split('|')
    const baseKeyString = parts[0]
    const transformerSpecs = parts.length > 1 ? parts.slice(1) : []

    const parser = new KeyParser()
    const key = parser.parse(baseKeyString, baseOffset)

    if (transformerSpecs.length > 0) {
        let runningOffset = baseOffset + baseKeyString.length + 1
        key.transformers = transformerSpecs.map((spec) => {
            const parsed = parseTransformerSpec(spec, runningOffset)
            if (!parsed.name) {
                throw new KeyParseError(
                    'empty transformer name in key',
                    new Range(runningOffset, runningOffset + spec.length),
                )
            }
            runningOffset += spec.length + 1
            return parsed
        })
        key.raw = keyString
        key.range = new Range(baseOffset, baseOffset + keyString.length)
    }

    return key
}
